cpp/src/io/utilities/data_casting.cu

```diff
@@ -0,0 +1,952 @@
+
+  // Copy literal/numeric value
+  if (not is_string_value) {
+    bytes += (in_end - in_begin);
+    if (d_buffer) d_buffer = thrust::copy(thrust::seq, in_begin, in_end, d_buffer);
+    return {bytes, data_casting_result::PARSING_SUCCESS};
+  }
+  char constexpr backslash_char = '\\';
+
+  // Escape-flag, set after encountering a backslash character
+  bool is_prev_char_escape = false;
+
+  // Exclude beginning and ending quote chars from string range
+  if (!options.keepquotes) {
+    ++in_begin;
+    --in_end;
+  }
+
+  // Iterate over the input
+  while (in_begin != in_end) {
+    // Copy single character to output
+    if (!is_prev_char_escape) {
+      is_prev_char_escape = (*in_begin == backslash_char);
+      if (!is_prev_char_escape) {
+        if (d_buffer) *d_buffer++ = *in_begin;
+        ++bytes;
+      }
+      ++in_begin;
+      continue;
+    }
+
+    // Previous char indicated beginning of escape sequence
+    // Reset escape flag for next loop iteration
+    is_prev_char_escape = false;
+
+    // Check the character that is supposed to be escaped
+    auto escaped_char = get_escape_char(*in_begin);
+
+    // We escaped an invalid escape character -> "fail"/null for this item
+    if (escaped_char == NON_ESCAPE_CHAR) { return {bytes, data_casting_result::PARSING_FAILURE}; }
+
+    // Regular, single-character escape
+    if (escaped_char != UNICODE_SEQ) {
+      if (d_buffer) *d_buffer++ = escaped_char;
+      ++bytes;
+      ++in_begin;
+      continue;
+    }
+
+    // This is an escape sequence of a unicode code point: \uXXXX,
+    // where each X in XXXX represents a hex digit
+    // Skip over the 'u' char from \uXXXX to the first hex digit
+    ++in_begin;
+
+    // Make sure that there's at least 4 characters left from the
+    // input, which are expected to be hex digits
+    if (thrust::distance(in_begin, in_end) < UNICODE_HEX_DIGIT_COUNT) {
+      return {bytes, data_casting_result::PARSING_FAILURE};
+    }
+
+    auto hex_val = parse_unicode_hex(in_begin);
+
+    // Couldn't parse hex values from the four-character sequence -> "fail"/null for this item
+    if (hex_val < 0) { return {bytes, data_casting_result::PARSING_FAILURE}; }
+
+    // Skip over the four hex digits
+    thrust::advance(in_begin, UNICODE_HEX_DIGIT_COUNT);
+
+    // If this may be a UTF-16 encoded surrogate pair:
+    // we expect another \uXXXX sequence
+    int32_t hex_low_val = 0;
+    if (thrust::distance(in_begin, in_end) >= NUM_UNICODE_ESC_SEQ_CHARS &&
+        *in_begin == backslash_char && *thrust::next(in_begin) == 'u') {
+      // Try to parse hex value following the '\' and 'u' characters from what may be a UTF16 low
+      // surrogate
+      hex_low_val = parse_unicode_hex(thrust::next(in_begin, 2));
+    }
+
+    // This is indeed a UTF16 surrogate pair
+    if (hex_val >= UTF16_HIGH_SURROGATE_BEGIN && hex_val < UTF16_HIGH_SURROGATE_END &&
+        hex_low_val >= UTF16_LOW_SURROGATE_BEGIN && hex_low_val < UTF16_LOW_SURROGATE_END) {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Is it valid to have two successive UTF-16 codepoints where the first is contained in `[HIGH_SURROGATE_BEGIN, HIGH_SURROGATE_END)` and the second is _not contained_ in `[LOW_SURROGATE_BEGIN, LOW_SURROGATE_END)`? If not, then I think we are missing a parse error here for this case.

Reply:
OK, so if I understand correctly, we want logic that is something like:

```
hex_val = parse_unicode_hex(stream);
if (is_high_surrogate(hex_val)) {
    auto hex_val_low = parse_unicode_hex(stream);
    if (is_low_surrogate(hex_val_low)) {
        // valid surrogate pair, decode
    } else {
        // invalid surrogate pair, parse error
        // I think this case is missing
        return {bytes, parse_error};
    }
} else {
   // non-surrogate pair, decode as normal
}
```

But I think the parse error in the case that we have a pair where the first is a surrogate pair and the second is not handled. Except that maybe this is deliberate per:

> The result is not valid UTF-16, but the majority of UTF-16 encoder and decoder implementations do this when translating between encodings.

So you're just "carrying on" and letting an invalid character appear in the output stream. Is that right?

If so, maybe

**suggestion:** Add a comment explaining this parsing behaviour.

Reply:
That's right.
added a comment at the place where this is not skipped.